### PR TITLE
Introducing stack-cmbm

### PR DESCRIPTION
**Pull request by baccenfutter**
This PR introduces:
- ies-rbenv
- stack-cmbm

The corresponding CMBM PR: https://github.com/easybib/citationmachine-bibme/pull/922

### CODE REVIEW

File: stack-cmbm/metadata.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name              'stack-cmbm'
+maintainer        'Till Klampaeckel'
```
Comment (till) [inline stack-cmbm/metadata.rb]: Maybe you want to change this? =)

File: stack-cmbm/recipes/deploy-ruby.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+end
+
+execute 'install ruby-2.2.3' do
+  command "ruby-install --system ruby #{ruby_version} -- --disable-install-rdoc"
+  not_if do
+    File.exist?('/usr/local/bin/ruby')
+  end
+end
+
+execute 'install bundler' do
```
Comment (till) [inline stack-cmbm/recipes/deploy-ruby.rb]: You could use `gem_package`.

Comment (till) [inline stack-cmbm/recipes/deploy-ruby.rb]: Also, generally related to the ruby setup. A `.gemrc` that disables installation of rdoc and tests buys you some time.

File: stack-cmbm/recipes/role-vagrant.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+required_packages = [
+  'libxml2', 'libmysqlclient-dev', 'nodejs', 'autoconf', 'bison', 'build-essential', 'libssl-dev', 'libyaml-dev',
```
Comment (till) [inline stack-cmbm/recipes/role-vagrant.rb]: We have a `nodejs` cookbook that's capable of installing specific versions.

See:
https://github.com/till/easybib-cookbooks/blob/master/stack-scholar/recipes/role-vagrant.rb#L9-L11

Configured via `web-dna.json` in Scholar's Vagrant (Line: 445-451)

File: stack-cmbm/recipes/role-vagrant.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  'libreadline6-dev', 'zlib1g-dev', 'libncurses5-dev', 'libffi-dev', 'libgdbm3', 'libgdbm-dev', 'libreadline-dev',
+  'redis-server', 'qt5-default', 'libqt5webkit5-dev', 'gstreamer1.0-plugins-base', 'gstreamer1.0-tools'
```
Comment (till) [inline stack-cmbm/recipes/role-vagrant.rb]: We also have one for `redis`.

**Comment by baccenfutter** [comment]
This PR still has an open dependency to https://github.com/easybib/citationmachine-bibme/pull/657

File: stack-cmbm/recipes/deploy-nginxapp.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
+include_recipe 'nginx-app::server'
+include_recipe 'supervisor'
+
+# Under which `user` should supervisor start and run the `puma` process.
+user = if is_aws
+         'www-data'
```
Comment (fh) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: Better use a node attr here - either `node['nginx-app']['user']`, or something new in the stack attribute rb

File: stack-cmbm/recipes/deploy-nginxapp.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+       else
+         'vagrant'
+       end
+
+node['vagrant']['applications'].each do |app_name, app_config|
```
Comment (fh) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: This is vagrant only?

File: stack-cmbm/recipes/deploy-nginxapp.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    domain_name domain_name
+    notifies :reload, 'service[nginx]', :delayed
+  end
+
+  easybib_envconfig app_name
+
+  supervisor_service 'puma_supervisor' do
+    action [:enable, :restart]
+    autostart true
+    command "puma -C #{app_dir}/config/puma.rb /vagrant_cmbm/config.ru"
```
Comment (fh) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: Why dont you use `#{app_dir}` for both?

File: stack-cmbm/recipes/deploy-vagrant.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+include_recipe 'nginx-app::server'
```
Comment (fh) [inline stack-cmbm/recipes/deploy-vagrant.rb]: Maybe add some fatal chef error on `if is_aws` to ensure this is only run on vagrant?

File: stack-cmbm/recipes/role-vagrant.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+# Configure nginx and supervisor
+include_recipe 'stack-cmbm::deploy-vagrant'
+
+execute 'install gems' do
+  command 'cd /vagrant_cmbm && bundle install'
+end
+
+execute 'setup app' do
+  command 'cd /vagrant_cmbm && bundle exec rake db:setup ads:populate'
```
Comment (fh) [inline stack-cmbm/recipes/role-vagrant.rb]: Maybe relocate those two execute statements to `deploy-vagrant` - this way, you dont have to hardcode the path, and it kind of is part of the deployment?

File: ies-ruby/providers/rubies.rb [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+def opsworks_mandatory_ruby
+  @opsworks_mandatory_ruby ||= new_resource.opsworks_mandatory_ruby
+end
+
+action :install do
+  desired_ruby = new_resource.desired_ruby
+
+  if Dir.exist?("#{rbenv_home}/versions/#{desired_ruby}")
+    Chef::Log.info("#{desired_ruby} is already installed, skipping!")
+  else
```
Comment (fh) [inline ies-ruby/providers/rubies.rb]: Please remove the else block, and use `next` in the if statement, makes it easier to read.

File: stack-cmbm/files/default/bashrc [hunk 11]
```diff
@@ -0,0 +1,102 @@
+# ~/.bashrc: executed by bash(1) for non-login shells.
```
Comment (fh) [inline stack-cmbm/files/default/bashrc]: where is this used? Also, maybe use `/etc/bashrc.d` instead, so you dont risk overwriting

File: stack-cmbm/files/default/zshrc [hunk 12]
```diff
@@ -0,0 +1,71 @@
+# Don't kill jobs when logout ---------------------------------------
```
Comment (fh) [inline stack-cmbm/files/default/zshrc]: Where is this used?

File: stack-cmbm/recipes/deploy-nginxapp.rb [hunk 13]
```diff
@@ -0,0 +1,101 @@
+include_recipe 'nginx-app::server'
```
Comment (fh) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: There is no deploy event of some sort in this recipe

File: stack-cmbm/templates/default/default-web-nginx.conf.erb [hunk 14]
```diff
@@ -0,0 +1,81 @@
+server {
+    listen 80 <%=@listen_opts%>;
+
+    index <%=@default_router%> index.php;
```
Comment (fh) [inline stack-cmbm/templates/default/default-web-nginx.conf.erb]: index.php is probably wrong in this stack

File: stack-cmbm/recipes/deploy-nginxapp.rb (hunk 13 above)
Comment (baccenfutter) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: Yes there is, it is hiding in line 60. Did you miss that or are you referring to something else?

Comment (baccenfutter) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: Granted. There is a major indention bug that very probably is somewhat confusing to the reader.

Comment (fh) [inline stack-cmbm/recipes/deploy-nginxapp.rb]: There is no `easybib_deploy` or something, just an nginx setup?

**Comment by baccenfutter** [comment]
Please review and merge.

File: stack-cmbm/templates/default/puma.conf.erb [hunk 15]
```diff
@@ -0,0 +1,3 @@
+<% @apps.each do |app| -%>
```
Comment (fh) [inline stack-cmbm/templates/default/puma.conf.erb]: @baccenfutter I guess this template can go? Nowhere used, and puma has been removed.

**Pull request merged**